runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java

```diff
@@ -448,18 +458,152 @@ public String toNativeString() {
     JavaRDD<KV<K, Iterable<WindowedValue<V>>>> groupRDD =
         GroupCombineFunctions.groupByKeyOnly(kvInRDD, keyCoder, wvCoder, partitioner);
 
-    return groupRDD
-        .map(
-            input -> {
-              final K key = input.getKey();
-              Iterable<WindowedValue<V>> value = input.getValue();
-              return FluentIterable.from(value)
-                  .transform(
-                      windowedValue ->
-                          windowedValue.withValue(KV.of(key, windowedValue.getValue())))
-                  .iterator();
-            })
-        .flatMapToPair(doFnFunction);
+    if (!requiresSortedInput) {
+      return groupRDD
+          .map(
+              input -> {
+                final K key = input.getKey();
+                Iterable<WindowedValue<V>> value = input.getValue();
+                return FluentIterable.from(value)
+                    .transform(
+                        windowedValue ->
+                            windowedValue.withValue(KV.of(key, windowedValue.getValue())))
+                    .iterator();
+              })
+          .flatMapToPair(doFnFunction);
+    }
+
+    JavaPairRDD<ByteArray, byte[]> pairRDD =
+        kvInRDD
+            .map(new ReifyTimestampsAndWindowsFunction<>())
+            .mapToPair(TranslationUtils.toPairFunction())
+            .mapToPair(
+                CoderHelpers.toByteFunctionWithTs(keyCoder, wvCoder, in -> in._2().getTimestamp()));
+
+    JavaPairRDD<ByteArray, byte[]> sorted =
+        pairRDD.repartitionAndSortWithinPartitions(keyPrefixPartitionerFrom(partitioner));
+
+    return sorted.mapPartitionsToPair(wrapDoFnFromSortedRDD(doFnFunction, keyCoder, wvCoder));
+  }
+
+  private static Partitioner keyPrefixPartitionerFrom(Partitioner partitioner) {
+    return new Partitioner() {
+      @Override
+      public int numPartitions() {
+        return partitioner.numPartitions();
+      }
+
+      @Override
+      public int getPartition(Object o) {
+        ByteArray b = (ByteArray) o;
+        return partitioner.getPartition(
+            new ByteArray(Arrays.copyOfRange(b.getValue(), 0, b.getValue().length - 8)));
```

Comment:
I wouldn't do that. The utility class is named `CoderHelpers` and this creates `Partitioner`, which has nothing to do with `Coders`. We would have to refactor it to something like `ByteArrayHelpers`, but I don't think that it would do this PR more readable. We might postpone that after time this is merged.